### PR TITLE
ADD GraphQL Client in frontend

### PR DESCRIPTION
**Pull request by pontushed**
Add ApolloClient
Template for queries
Template in app.js for login, token etc

### CODE REVIEW

File: frontend/src/App.js [hunk 1]
```diff
@@ -2,23 +2,46 @@ import React, { useState, useEffect } from 'react'
...
+** Remove comments when ready to implement
+
+import {
+  useQuery,
+  useApolloClient,
+  useMutation,
+  useSubscription,
+} from '@apollo/client'
+import { LOGIN } from './graphql/queries'
+*/
```
Comment (pontushed) [inline frontend/src/App.js:16]: Uncomment when ready to implement

File: frontend/src/App.js [hunk 2]
```diff
@@ -2,23 +2,46 @@ import React, { useState, useEffect } from 'react'
...
 
 const App = () => {
 
   const [events, setEvents] = useState([])
-
+  // GraphQL Client variables
+  // const client = useApolloClient()
+  // const [token, setToken] = useState(null)
+  // END GraphQL Client variables
```
Comment (pontushed) [inline frontend/src/App.js:24]: Uncomment when ready to implement

File: frontend/src/App.js [hunk 3]
```diff
@@ -2,23 +2,46 @@ import React, { useState, useEffect } from 'react'
...
+  /* USE TOKEN IF EXISTS
+  useEffect(() => {
+    const token = localStorage.getItem('libraryapp-token')
+    if (token) {
+      setToken(token)
+    }
+  }, [])
+
+  */
+
```
Comment (pontushed) [inline frontend/src/App.js:39]: If token exists in LocalStorage, use it in the headers for the queries. Btw, change the item name from 'libraryapp-token' to something else!

File: frontend/src/App.js [hunk 4]
```diff
@@ -2,23 +2,46 @@ import React, { useState, useEffect } from 'react'
...
   return (
     <div className="App">
       <MyCalendar events={events}/>
     </div>
   )
 }
 
-// MUN FEATURE
-
```
Comment (pontushed) [inline frontend/src/App.js:23]: Clean up

File: frontend/src/graphql/queries.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import { gql } from '@apollo/client'
+
+export const LOGIN = gql`
+  mutation login($username: String!, $password: String!) {
+    login(username: $username, password: $password) {
+      value
+    }
+  }
+`
+
```
Comment (pontushed) [inline frontend/src/graphql/queries.js:10]: Basic setup. Additional queries added here.

File: frontend/src/index.js [hunk 6]
```diff
@@ -1,9 +1,38 @@
 import React from 'react'
 import ReactDOM from 'react-dom'
+import { setContext } from 'apollo-link-context'
```
Comment (pontushed) [inline frontend/src/index.js:3]: Used for passing auth header

File: frontend/src/index.js [hunk 7]
```diff
@@ -1,9 +1,38 @@
...
+
+import {
+  ApolloClient,
+  ApolloProvider,
+  HttpLink,
+  InMemoryCache,
+} from '@apollo/client'
+
```
Comment (pontushed) [inline frontend/src/index.js:11]: Necessary parts for the client

File: frontend/src/index.js [hunk 8]
```diff
@@ -1,9 +1,38 @@
...
+const authLink = setContext((_, { headers }) => {
+  const token = localStorage.getItem('libraryapp-token')
+
+  return {
+    headers: {
+      ...headers,
+      authorization: token ? `bearer ${token}` : null,
+    },
+  }
+})
```
Comment (pontushed) [inline frontend/src/index.js:24]: Part for passing the auth header in the queries

File: frontend/src/index.js [hunk 9]
```diff
@@ -1,9 +1,38 @@
...
+
+const httpLink = new HttpLink({ uri: 'http://localhost:3001' })
```
Comment (pontushed) [inline frontend/src/index.js:26]: Main link for queries. Change to match backend, if necessary (Probably when going to production)

File: frontend/src/index.js [hunk 10]
```diff
@@ -1,9 +1,38 @@
...
+
+const client = new ApolloClient({
+  cache: new InMemoryCache(),
+  link: authLink.concat(httpLink),
+})
+
```
Comment (pontushed) [inline frontend/src/index.js:32]: Make client using InMemoryCache and combine authLink and httpLink

**Comment by tsa-dom** [comment]
Vaikuttaa hyvältä!

File: frontend/src/index.js [hunk 11]
```diff
@@ -1,9 +1,38 @@
...
 ReactDOM.render(
-  <App />,
+  <ApolloProvider client={client}>
+    <App />
+  </ApolloProvider>,
```
Comment (pontushed) [inline frontend/src/index.js:36]: Wrap App as a child of ApolloProvider, so the client can be used in all components

**Pull request merged**